File: db4on/test/src/com/db4o/db4ounit/common/assorted/MultiDeleteTestCase.cs
namespace com.db4o.db4ounit.common.assorted
{
	public class MultiDeleteTestCase : Db4oUnit.Extensions.AbstractDb4oTestCase, Db4oUnit.Extensions.Fixtures.OptOutDefragSolo
	{
		public static void Main(string[] args)
		{
			new com.db4o.db4ounit.common.assorted.MultiDeleteTestCase().RunSoloAndClientServer
				();
		}

		public class Item
		{
			public com.db4o.db4ounit.common.assorted.MultiDeleteTestCase.Item child;

			public string name;

			public object forLong;

			public long myLong;

			public object[] untypedArr;

			public long[] typedArr;

			public virtual void SetMembers()
			{
				forLong = System.Convert.ToInt64(100);
				myLong = System.Convert.ToInt64(100);
				untypedArr = new object[] { System.Convert.ToInt64(10), "hi", new com.db4o.db4ounit.common.assorted.MultiDeleteTestCase.Item
					() };
				typedArr = new long[] { System.Convert.ToInt64(3), System.Convert.ToInt64(7), System.Convert.ToInt64
					(9) };
			}
		}

		protected override void Configure(com.db4o.config.Configuration config)
		{
			com.db4o.config.ObjectClass itemClass = config.ObjectClass(typeof(com.db4o.db4ounit.common.assorted.MultiDeleteTestCase.Item)
				);
			itemClass.CascadeOnDelete(true);
			itemClass.CascadeOnUpdate(true);
		}

		protected override void Store()
		{
			com.db4o.db4ounit.common.assorted.MultiDeleteTestCase.Item md = new com.db4o.db4ounit.common.assorted.MultiDeleteTestCase.Item
				();
			md.name = "killmefirst";
			md.SetMembers();
			md.child = new com.db4o.db4ounit.common.assorted.MultiDeleteTestCase.Item();
			md.child.SetMembers();
			Db().Set(md);
		}

		public virtual void TestDeleteCanBeCalledTwice()
		{
			com.db4o.db4ounit.common.assorted.MultiDeleteTestCase.Item item = ItemByName("killmefirst"
				);
			Db4oUnit.Assert.IsNotNull(item);
			long id = Db().GetID(item);
			Db().Delete(item);
			Db4oUnit.Assert.AreSame(item, ItemById(id));
			Db().Delete(item);
			Db4oUnit.Assert.AreSame(item, ItemById(id));
		}

		private com.db4o.db4ounit.common.assorted.MultiDeleteTestCase.Item ItemByName(string
			 name)
		{
			com.db4o.query.Query q = NewQuery(typeof(com.db4o.db4ounit.common.assorted.MultiDeleteTestCase.Item)
				);
			q.Descend("name").Constrain(name);
			return (com.db4o.db4ounit.common.assorted.MultiDeleteTestCase.Item)q.Execute().Next
				();
		}

		private com.db4o.db4ounit.common.assorted.MultiDeleteTestCase.Item ItemById(long 
			id)
		{
			return (com.db4o.db4ounit.common.assorted.MultiDeleteTestCase.Item)Db().GetByID(id
				);
		}
	}
}

File: db4ocs/test/src/com/db4o/test/MultiDeleteTestCase.java
/* Copyright (C) 2004 - 2007  db4objects Inc.  http://www.db4o.com */

package com.db4o.test;

import com.db4o.*;
import com.db4o.config.*;
import com.db4o.ext.*;
import com.db4o.query.*;

import db4ounit.*;
import db4ounit.extensions.*;

public class MultiDeleteTestCase extends Db4oClientServerTestCase {

	public static void main(String[] args) {
		new MultiDeleteTestCase().runConcurrency();
	}
	
	MultiDeleteTestCase child;

	String name;

	Object forLong;

	Long myLong;

	Object[] untypedArr;

	Long[] typedArr;

	public void configure(Configuration config) {
		config.objectClass(this).cascadeOnDelete(true);
		config.objectClass(this).cascadeOnUpdate(true);
	}

	public void store() {
		MultiDeleteTestCase md = new MultiDeleteTestCase();
		md.name = "killmefirst";
		md.setMembers();
		md.child = new MultiDeleteTestCase();
		md.child.setMembers();
		store(md);
	}

	public void conc(ExtObjectContainer oc) throws Exception {
		Query q = oc.query();
		q.constrain(MultiDeleteTestCase.class);
		q.descend("name").constrain("killmefirst");
		ObjectSet objectSet = q.execute();
		if (objectSet.size() == 0) { // already deleted by other threads
			return;
		}
		
		Assert.areEqual(1, objectSet.size());
		Thread.sleep(1000);
		MultiDeleteTestCase md = (MultiDeleteTestCase) objectSet.next();
		oc.delete(md);
		assertCountOccurences(oc, MultiDeleteTestCase.class, 0);
	}

	public void check(ExtObjectContainer oc) {
		assertCountOccurences(oc, MultiDeleteTestCase.class, 0);
	}

	private void setMembers() {
		forLong = new Long(100);
		myLong = new Long(100);
		untypedArr = new Object[] { new Long(10), "hi", new MultiDeleteTestCase() };
		typedArr = new Long[] { new Long(3), new Long(7), new Long(9), };
	}

}
